File: progred_graph/src/gid.rs
use crate::id::Id;
use im::HashMap;

pub trait Gid {
    fn edges(&self, entity: &Id) -> Option<&HashMap<Id, Id>>;

    fn get(&self, entity: &Id, label: &Id) -> Option<&Id> {
        self.edges(entity)?.get(label)
    }
}

pub struct StackedGid<'a, Top: Gid, Bottom: Gid> {
    top: &'a Top,
    bottom: &'a Bottom,
}

impl<'a, Top: Gid, Bottom: Gid> StackedGid<'a, Top, Bottom> {
    pub fn new(top: &'a Top, bottom: &'a Bottom) -> Self {
        Self { top, bottom }
    }
}

impl<Top: Gid, Bottom: Gid> Gid for StackedGid<'_, Top, Bottom> {
    fn edges(&self, entity: &Id) -> Option<&HashMap<Id, Id>> {
        self.top.edges(entity).or_else(|| self.bottom.edges(entity))
    }
}

File: progred/progred/Gid.swift
import Foundation
import HashTreeCollections

protocol Gid {
    func edges(entity: Id) -> TreeDictionary<Id, Id>?

    func get(entity: Id, label: Id) -> Id?
}

extension Gid {
    func get(entity: Id, label: Id) -> Id? {
        edges(entity: entity)?[label]
    }
}

struct StackedGid<Top: Gid, Bottom: Gid>: Gid {
    let top: Top
    let bottom: Bottom

    func edges(entity: Id) -> TreeDictionary<Id, Id>? {
        top.edges(entity: entity) ?? bottom.edges(entity: entity)
    }
}
